hazelcast/src/main/java/com/hazelcast/sql/impl/operation/QueryExecuteOperationFactory.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.operation;
+
+import com.hazelcast.sql.impl.QueryId;
+import com.hazelcast.sql.impl.plan.Plan;
+import com.hazelcast.sql.impl.plan.PlanFragmentMapping;
+import com.hazelcast.sql.impl.plan.node.PlanNode;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import static com.hazelcast.sql.impl.operation.QueryExecuteOperationFragmentMapping.DATA_MEMBERS;
+import static com.hazelcast.sql.impl.operation.QueryExecuteOperationFragmentMapping.EXPLICIT;
+
+/**
+ * Factory to create query execute operations.
+ */
+public class QueryExecuteOperationFactory {
+
+    private final Plan plan;
+    private final List<Object> args;
+    private final Map<Integer, Long> edgeInitialMemoryMap;
+
+    public QueryExecuteOperationFactory(
+        Plan plan,
+        List<Object> args,
+        Map<Integer, Long> edgeInitialMemoryMap
+    ) {
+        this.plan = plan;
+        this.args = args;
+        this.edgeInitialMemoryMap = edgeInitialMemoryMap;
+    }
+
+    public QueryExecuteOperation create(QueryId queryId, UUID targetMemberId) {
+        int fragmentCount = plan.getFragmentCount();
+
+        // Prepare descriptors.
+        List<QueryExecuteOperationFragment> fragments = new ArrayList<>(fragmentCount);
+
+        for (int i = 0; i < fragmentCount; i++) {
+            PlanFragmentMapping planMapping = plan.getFragmentMapping(i);
+
+            QueryExecuteOperationFragmentMapping mapping;
+            Collection<UUID> memberIds;
+            PlanNode node;
+
+            if (planMapping.isDataMembers()) {
+                mapping = DATA_MEMBERS;
+                memberIds = null;
+                node = plan.getFragment(i);
+            } else {
+                mapping = EXPLICIT;
+                memberIds = planMapping.getMemberIds();
+                node = memberIds.contains(targetMemberId) ? plan.getFragment(i) : null;
```

Comment:
Not sure, what condition absence of a member in the member list indicates?

Reply:
The `EXPLICIT` mapping means that we defined the list of nodes where a fragment should be executed. If the member is not in the list, then the fragment should not be executed on it. We indicate the condition "do not execute that fragment on that node" via `node == null`.

Reply:
In 4.1 we will have only one example of such a fragment - `root` fragment.

Reply:
Probably I'm getting something wrong, but the only thing we do with fragments having `null` plan is skipping them: https://github.com/hazelcast/hazelcast/pull/16876/files#diff-a72004010c5a07d55b4d36beeb25828bR184, should we just skip their creation?

Reply:
The fragment is defined by a node to be executed and target members. In order to set up streams between send and receive fragments, we need to know the number of inputs and outputs for every edge. E.g. a `Receiver[member1, member2]` must know that it communicates with `Sender[member1, member2, member3]`, and therefore it should wait for 3 end markers. So we use the `QueryExecuteOperationFragment` not only to pass the `PlanNode` but mapping information as well, that is needed by IO nodes.

You may see it in action in the usages of `CreateExecPlanNodeVisitor.getFragmentMembers` - sender gets information about receiver and vice versa.

Reply:
I see, probably that worth a mention somewhere in `QueryExecuteOperationFragment` javadoc, so one can easily understand that fragments having nothing to execute (`PlanNode` is `null`) are valid and what their purpose is.

Reply:
Added JavaDoc.